File: src/main/java/com/ndgndg91/ordermatchingengine/order/service/OrderBooks.java
package com.ndgndg91.ordermatchingengine.order.service;

import com.ndgndg91.ordermatchingengine.global.OrderServiceException;
import com.ndgndg91.ordermatchingengine.order.*;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

@Service
public class OrderBooks {
    private static final String MESSAGE = "Not Found Symbol %s";
    private final Map<Symbol, OrderBook> m;

    public OrderBooks() {
        this.m = new ConcurrentHashMap<>();
        for (var s : Symbol.values()) {
            this.m.put(s, new OrderBook(s));
        }
    }

    public void addOrder(Order order) {
        OrderBook orderBook = m.get(order.getSymbol());
        if (orderBook == null)
            throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), String.format(MESSAGE, order.getSymbol()));
        orderBook.addOrder(order);
    }

    public void modifyOrder(Order order) {
        OrderBook orderBook = m.get(order.getSymbol());
        if (orderBook == null)
            throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), String.format(MESSAGE, order.getSymbol()));
        orderBook.modifyOrder(order);
    }

    public void cancelOrder(Order order) {
        OrderBook orderBook = m.get(order.getSymbol());
        if (orderBook == null)
            throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), String.format(MESSAGE, order.getSymbol()));
        orderBook.cancelOrder(order);
    }

    public Optional<MatchResult> match(Symbol symbol, PriceType priceType, OrderType orderType) {
        OrderBook orderBook = m.get(symbol);
        if (orderBook == null)
            throw new OrderServiceException(null, HttpStatus.BAD_REQUEST.value(), String.format(MESSAGE, symbol));
        return Optional.ofNullable(orderBook.match(priceType, orderType));
    }

    public Optional<OrderBook> findOrderBooksBySymbol(Symbol symbol) {
        return Optional.ofNullable(m.get(symbol));
    }
}

File: OrderMatchingEngineKotlin/src/main/kotlin/com/ndgndg91/ordermatchingenginekotlin/order/service/OrderBooks.kt
package com.ndgndg91.ordermatchingenginekotlin.order.service

import com.ndgndg91.ordermatchingenginekotlin.global.OrderServiceException
import com.ndgndg91.ordermatchingenginekotlin.order.*
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import java.util.concurrent.ConcurrentHashMap

@Service
class OrderBooks {
    companion object {
        const val MESSAGE = "Not Found Symbol : %s \t OrderId : %s"
    }

    private val m: ConcurrentHashMap<Symbol, OrderBook> = ConcurrentHashMap(Symbol.values().associateWith { OrderBook(it) })

    fun addOrder(order: Order) {
        val orderBook = m[order.symbol]
        if (kotlin.runCatching { orderBook!!.addOrder(order) }.isFailure) {
            throw OrderServiceException(
                null,
                HttpStatus.BAD_REQUEST.value(),
                String.format(MESSAGE, order.symbol, order.orderId)
            )
        }
    }

    fun modifyOrder(order: Order) {
        val orderBook = m[order.symbol]
        if (kotlin.runCatching { orderBook!!.modifyOrder(order) }.isFailure) {
            throw OrderServiceException(
                null,
                HttpStatus.BAD_REQUEST.value(),
                String.format(MESSAGE, order.symbol, order.orderId)
            )
        }
    }

    fun cancelOrder(order: Order) {
        val orderBook = m[order.symbol]
        if (kotlin.runCatching { orderBook!!.cancelOrder(order) }.isFailure) {
            throw OrderServiceException(
                null,
                HttpStatus.BAD_REQUEST.value(),
                String.format(MESSAGE, order.symbol, order.orderId)
            )
        }
    }

    fun find(symbol: Symbol, orderType: OrderType, orderId: String): OrderEntry? {
        val orderBook = m[symbol]
        val findResult = kotlin.runCatching { orderBook!!.find(orderType, orderId) }
        if (findResult.isFailure) {
            throw OrderServiceException(
                null,
                HttpStatus.BAD_REQUEST.value(),
                String.format(MESSAGE, symbol, orderId)
            )
        }

        return findResult.getOrNull()
    }

    fun match(symbol: Symbol, priceType: PriceType, orderType: OrderType): List<MatchResult> {
        val orderBook = m[symbol]
        return orderBook!!.match(priceType, orderType)
    }
}